hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/states/NodeStateMap.java

```diff
@@ -224,6 +224,23 @@ public NodeState getNodeState(UUID uuid) throws NodeNotFoundException {
     }
   }
 
+  /**
+   * Adds the given container to the specified datanode.
+   *
+   * @param uuid - datanode uuid
+   * @param containerId - containerID
+   * @throws NodeNotFoundException - if datanode is not known. For new datanode
+   *                        use addDatanodeInContainerMap call.
+   */
+  public void addContainer(final UUID uuid,
+                           final ContainerID containerId)
+      throws NodeNotFoundException {
+    if (!nodeToContainer.containsKey(uuid)) {
+      throw new NodeNotFoundException("Node UUID: " + uuid);
+    }
+    nodeToContainer.get(uuid).add(containerId);
```

Comment:
the check and get calls can race with the remove call.

Also looking into the code, there is no remove call, we should remove the dn once it is dead ?

Reply:
For now we don't ever remove a datanode from NodeManager once it is registered. We should not remove dn once it is dead, we will not be able to show list of dead nodes to the user.
Maybe as part of decommissioning once the decommission is successful we can remove the dn.